### ISSUE TITLE
@page for any file

### ISSUE THREAD

**Issue by mokone91** [KnodesCommunity/typedoc-plugins#132]
Hi! 
Is its possible to add a link to page for not described pages in config?

example: 
I want to past the link to my change-log file at the bottom of readme, i have around 40 packages and all of theme have readme and changelog files at the root. i do not want to discribe the whole pages tree..

**Comment by GerkinDev** [comment]
Hi there,

Well, I have more or less planned to add the concept of _front matter_ similar to what can be done with [some static side generators like _hugo_](https://gohugo.io/content-management/front-matter/).

With such feature, you could specify a glob pattern to match pages, and specify the page infos (such as title or output) directly in the front matter.
Menu/directory configuration (such as children output directory or upcoming options evoked in #100) might then be done in a config file in each directory.

If you have any proposal/suggestion, I'm all ears.

Yet, what you want requires a full rework of the pages building process, and is not really a top priority.

**Comment by GerkinDev** [comment]
I just thought that changelogs may be auto-generated & published in other tools (GitHub releases, npm tarball, whatever), and thus, adding a front-matter does not really fit.

In your case, I would think of something like: 
```js
{
    pages: [
        {match: './packages/*/CHANGELOG.md', template: {
            moduleRoot: true,
            title: "{{CURRENT_MODULE.NAME}}",
            children: [
                {source: "./CHANGELOG.md",title: "Changelog"}
            ]
        }}
    ]
}
```

This would be expanded as following:
```js

{
    pages: [
        {
            moduleRoot: true,
            title: "@example/foo",
            children: [
                {source: "./CHANGELOG.md",title: "Changelog"}
            ]
        },
        {
            moduleRoot: true,
            title: "@example/bar",
            children: [
                {source: "./CHANGELOG.md",title: "Changelog"}
            ]
        },
    ]
}
```

---

You could even append README to every modules with such pattern, making [@knodes/typedoc-plugin-monorepo-readmes](https://www.npmjs.com/package/@knodes/typedoc-plugin-monorepo-readmes) redundant:

```js
{
    pages: [
        {match: './packages/*/README.md', template: {
            moduleRoot: true,
            title: "{{CURRENT_MODULE.NAME}}",
            source: "./README.md"
        }}
    ]
}
```

What do you think ?

**Comment by mokone91** [comment]
Yeah! Perfect for me!) is there a way to include index inside module lvl pages tree? 
<img width="323" alt="image" src="https://user-images.githubusercontent.com/2384641/179984416-e7280d42-f9ab-4387-98c9-6af4639ac2d8.png">

**Comment by GerkinDev** [comment]
Could you please clarify what you want to do ?

**Comment by mokone91** [comment]
Im happy to use glob pattern in root config when be available!

It will be nice if i can include index block directly to module subtree, but i understand that it not related to current topic and pages plug-in.
I subscribed to https://github.com/KnodesCommunity/typedoc-plugins/issues/109

Thanks!

**Comment by mokone91** [comment]
But, if we will use pages for readme - what we will se in module homepage?

**Comment by GerkinDev** [comment]
As I mentioned above, glob patterns would be expanded to the non-glob format:
```js
        {match: './packages/*/README.md', template: {
            moduleRoot: true,
            title: "{{CURRENT_MODULE.NAME}}",
            source: "./README.md"
            children: [
              { source: './CHAMGELOG.md', title: 'Changelog'}
           ]
        }}
```

This would expand to 
```
Module1 -> shows readme in Module1 + normal index
 - Changelog -> shows changelog of Module1
 - ...
Module2 -> shows readme in Module2 + normal index
 - Changelog -> shows changelog of Module2
 - ...
```

Would that matches your expectations ?
There would be an extra behavior, that merges deeply items, so that you could specify pages for modules both via glob expansion & explicitly like now

**Comment by mokone91** [comment]
Ahh, it's can be managed using  `moduleRoot: true`, sorry miss that.

Yeah perfect! Thats exactly what I need.

**Comment by GerkinDev** [comment]
Yeah this option was introduced in `0.23.0`, see [the docs page](https://knodescommunity.github.io/typedoc-plugins/_knodes_typedoc_plugin_pages/pages/pages-tree.html#monorepos).

Don't hesitate to suggest improvements to documentation

**Comment by mokone91** [comment]
Hi! can we also add ability to use fileName for page title somehow?
So it can be something like
```
{
  match: './packages/*/README.md',
  template: {
    moduleRoot: true,
    title: "{{CURRENT_MODULE.NAME}}",
    source: "./README.md",
    children: [
      { source: './CHAMGELOG.md', title: 'Changelog'},
      { source: './pages/*.md', title: "{{FILENAME}}"}
   ]
}}
```

Not sure if its possible, but i want to achieve the ability to add any pages to any modules without adding extra configuration rows to root config

Or maybe we can allow to add config for module lvl config? something like https://typedoc.org/guides/monorepo/#typedoc-package-config

Thanks!

**Comment by GerkinDev** [comment]
Hi!

So, about the 1st point, implementation is mostly done and you'll be able to interpolate almost any field from the path matched, using [lodash templates](https://lodash.com/docs/#template) in which you'll be able to use various fields from matches (including parents in case of nested templates), and use any lodash methods and a couple of cherry picked functions such as `path`'s `basename`, `dirname` and a few others.
The only thing I haven't implemented yet is using functions in place of template, so that you could specify advanced logic in your typedoc config as long as it is a js file.
Introducing this feature have caveats tough, because some options defaults cause quite counter intuitive behaviors. I'm still working on the backward compatibility.

About the config splitted across packages, you can already somewhat emulate that by requiring modules configs from your root typedoc.js and do your stuff there. Built in support of composite config is not that trivial when you check how typedoc works, but I'm thinking about it.

**Comment by mokone91** [comment]
Sounds great! I'm waiting impatiently)
Regarding requiring modules configs - yeah this should work, then we need to modify root config each time when we add new package... I believe templates feature will solve my needs)

**Comment by mokone91** [comment]
Have modified your snippet to attach chengelogs if exists
```
const pagesTree = readdirSync('packages').reduce((acc, pkg) => {
  if (existsSync(`./packages/${pkg}/CHANGELOG.md`)) {
    const pkgName = require(`./packages/${pkg}/package.json`).name;
    acc.push({
      title: pkgName,
      moduleRoot: true,
      childrenSourceDir: '.',
      children: [
        {
          title: 'Changelog',
          source: 'CHANGELOG.md',
        },
      ],
    });
  }
  return acc;
}, []);
```

but this works only if i place CHANGELOG.md to 'pages' folder inside package, even when 'childrenSourceDir': '.'

My changelog file placed directly at package root..

Seems i miss something..

**Comment by GerkinDev** [comment]
Yes this is due to an option I'm about to deprecate. Please set the [source option to `""`](https://knodescommunity.github.io/typedoc-plugins/_knodes_typedoc_plugin_pages/pages/options.html)

**Comment by mokone91** [comment]
yep, adding `source: ''` to root pages config make it work! thanks!

**Comment by GerkinDev** [comment]
Hi,

So, I thought I was pretty done with template nodes. But after updating the docs, I realized that the result was super verbose [as you can see in this repository's configuration] (https://github.com/KnodesCommunity/typedoc-plugins/blob/ec5d3bdfea291cad9939a97c5643b979281f5d0e/typedoc.js).
You can see the draft documentation [here](https://github.com/KnodesCommunity/typedoc-plugins/blob/ec5d3bdfea291cad9939a97c5643b979281f5d0e/packages/plugin-pages/pages/pages-tree.md).

So I decided to made the integration of this feature more SOLID by normalizing interfaces & facilitate further extensions.

That's why I decided to add the front-matter idea in the branch at the same time.

It also gives me a nice occasion to think about how to inject theme-specific options along with pages. 

I hope you'll understand and be patient.
Cheers!

**Comment by GerkinDev** [comment]
Hey there,

Could you check out latest `next` version ? https://www.npmjs.com/package/@knodes/typedoc-plugin-pages/v/0.23.5-next.1

You'll find the documentation for the new loader I hope you'll like here: https://knodescommunity.github.io/typedoc-plugins/v0.23.5-next.1/pages/_knodes_typedoc_plugin_pages/managing-pages-menu-items/using-the-front-matter-loader.html

The documentation is still a bit lacking, but I'm working on it.
